File: python-flask/test_api.py
# Runs an in-process smoke test against the rebuilt flat-file Flask RedirectIQ app.
import os
import sys
import time
import uuid

sys.path.insert(0, os.path.dirname(__file__))

from app import create_app


def main():
    app = create_app()
    client = app.test_client()
    email = f"redirectiq-smoke-{uuid.uuid4().hex[:8]}@test.com"
    custom_slug = f"smoke{uuid.uuid4().hex[:6]}"
    credentials = {
        "email": email,
        "password": "password123",
    }

    print("--- TEST 1: Register ---")
    register_response = client.post("/auth/register", json=credentials)
    print("Status:", register_response.status_code)
    print("Data:", register_response.get_json())

    print("--- TEST 2: Login ---")
    login_response = client.post("/auth/login", json=credentials)
    print("Status:", login_response.status_code)
    print("Data:", login_response.get_json())
    token = (login_response.get_json() or {}).get("token")

    if not token:
        raise RuntimeError("No token returned from login")

    headers = {"Authorization": f"Bearer {token}"}

    print("--- TEST 3: Create Link ---")
    create_response = client.post(
        "/links",
        json={"original_url": "https://google.com", "custom_slug": custom_slug},
        headers=headers,
    )
    print("Status:", create_response.status_code)
    print("Data:", create_response.get_json())
    created_link = create_response.get_json() or {}
    link_id = created_link.get("id")
    slug = created_link.get("slug")

    if not link_id or not slug:
        raise RuntimeError("No link id or slug returned from link creation")

    print("--- TEST 4: List Links ---")
    list_response = client.get("/links", headers=headers)
    print("Status:", list_response.status_code)
    print("Data:", list_response.get_json())

    print("--- TEST 5: Redirect By Slug ---")
    redirect_response = client.get(f"/{slug}", follow_redirects=False)
    print("Status:", redirect_response.status_code)
    print("Location:", redirect_response.headers.get("Location"))

    print("--- TEST 6: Stats Summary ---")
    summary_response = None

    for _ in range(10):
        summary_response = client.get("/stats/summary", headers=headers)
        payload = summary_response.get_json() or {}

        if payload.get("totalClicks", 0) >= 1:
            break

        time.sleep(0.1)

    print("Status:", summary_response.status_code)
    print("Data:", summary_response.get_json())

    print("--- TEST 7: Link Stats ---")
    stats_response = client.get(f"/links/{link_id}/stats", headers=headers)
    print("Status:", stats_response.status_code)
    print("Data:", stats_response.get_json())

    print("ALL TESTS DONE")


if __name__ == "__main__":
    main()

File: node-express/test-api.js
// Runs a sequential axios-based smoke test against the local RedirectIQ API.
const axios = require('axios');
const { once } = require('events');
const { startServer } = require('./src/server');

const HOST = '127.0.0.1';
const DEFAULT_PASSWORD = 'password123';

function createClient(baseUrl) {
  return axios.create({
    baseURL: baseUrl,
    validateStatus: function validateStatus() {
      return true;
    },
    maxRedirects: 0
  });
}

function logResponse(response) {
  console.log('Status:', response.status);
  console.log('Data:', response.data);
}

async function resolveBaseUrl() {
  const configuredBaseUrl = (process.env.BASE_URL || '').trim().replace(/\/$/, '');

  if (configuredBaseUrl) {
    return {
      baseUrl: configuredBaseUrl,
      close: async function closeConfiguredBaseUrl() {}
    };
  }

  const server = startServer({
    host: HOST,
    port: 0
  });

  await once(server, 'listening');

  const address = server.address();
  const baseUrl = `http://${HOST}:${address.port}`;

  return {
    baseUrl,
    close: async function closeStartedServer() {
      await new Promise(function closeServer(resolve, reject) {
        server.close(function onClose(error) {
          if (error) {
            reject(error);
            return;
          }

          resolve();
        });
      });
    }
  };
}

async function runTests() {
  const email = `redirectiq-smoke-${Date.now()}@test.com`;
  const customSlug = `smoke${Date.now().toString().slice(-6)}`;
  const credentials = {
    email,
    password: DEFAULT_PASSWORD
  };
  const { baseUrl, close } = await resolveBaseUrl();
  const client = createClient(baseUrl);

  try {
    let token;
    let slug;

    console.log(`Running smoke test against ${baseUrl}`);

    console.log('--- TEST 1: Register ---');
    const registerResponse = await client.post('/auth/register', credentials);
    logResponse(registerResponse);

    console.log('--- TEST 2: Login ---');
    const loginResponse = await client.post('/auth/login', credentials);
    logResponse(loginResponse);
    token = loginResponse.data && loginResponse.data.token;

    if (!token) {
      throw new Error('No token returned from login');
    }

    console.log('--- TEST 3: Create Link ---');
    const createLinkResponse = await client.post(
      '/links',
      {
        original_url: 'https://google.com',
        custom_slug: customSlug
      },
      {
        headers: {
          Authorization: `Bearer ${token}`
        }
      }
    );
    logResponse(createLinkResponse);
    slug = createLinkResponse.data && createLinkResponse.data.slug;

    if (!slug) {
      throw new Error('No slug returned from link creation');
    }

    console.log('--- TEST 4: List Links ---');
    const listLinksResponse = await client.get('/links', {
      headers: {
        Authorization: `Bearer ${token}`
      }
    });
    logResponse(listLinksResponse);

    console.log('--- TEST 5: Redirect By Slug ---');
    const redirectResponse = await client.get(`/${slug}`);
    console.log('Status:', redirectResponse.status);
    console.log('Location:', redirectResponse.headers.location || null);

    console.log('--- TEST 6: Stats Summary ---');
    const statsResponse = await client.get('/stats/summary', {
      headers: {
        Authorization: `Bearer ${token}`
      }
    });
    logResponse(statsResponse);
  } finally {
    await close();
  }
}

runTests()
  .catch(function handleError(error) {
    if (error.response) {
      console.error('Request failed with response:');
      console.error('Status:', error.response.status);
      console.error('Data:', error.response.data);
      return;
    }

    console.error('Test run failed:', error && (error.stack || error.message || error));
  })
  .finally(function finish() {
    console.log('ALL TESTS DONE');
  });
